### PR TITLE
Feature/#7/new scene 8 robots

### PR DESCRIPTION
**Pull request by 1475963**
* add 8 bots scene
* comment Population.cpp
* merge code refactoring updates
* adapt code to multiple robots (Simulation.cpp and Robot.cpp), somehow rdy for parallelism, currently set to 4 robots

### CODE REVIEW

File: srcs/Robot.cpp [hunk 1]
```diff
@@ -45,9 +45,15 @@ Robot::Robot(simxInt clientID) :
   _clientID(clientID), _handlers(highestAction()), _id(_nextId) {
   ++_nextId;
   // here we throw but it's not a problem as this will later be done in another function
-  for (int i = 0; i < highestAction(); i++)
-    if (simxGetObjectHandle(_clientID, Robot::actions.at(i)._bodyPart, &_handlers[i], simx_opmode_blocking) != 0)
+  const char *id = std::to_string(_id).c_str();
+  for (int i = 0; i < highestAction(); i++) {
+    char *completeBodyPart = new char[strlen(Robot::actions.at(i)._bodyPart) + strlen(id) + 1]();
```
Comment (JeremyPouyet) [inline srcs/Robot.cpp]: can't lines 50 to 53 be replaced with the use of std::string ?

File: includes/Robot.hh [hunk 2]
```diff
@@ -5,6 +5,9 @@
 #include <iostream> // cout / cerr
 #include <unistd.h> // sleep
 
+#define _USE_MATH_DEFINES // pi constant
+#include <math.h> // pi constant
```
Comment (JeremyPouyet) [inline includes/Robot.hh]: in c++ you usually include <cmath> and not <math.h>

File: srcs/Simulation.cpp [hunk 3]
```diff
@@ -2,18 +2,25 @@
...
   _population(Population(_maxPop)),
-  _clientID(simxStart("127.0.0.1", 19997, true, true, 5000, 5)),
-  _robot(Robot(_clientID)) {
+  _clientID(simxStart("127.0.0.1", 19997, true, true, 5000, 5)) {
   if (_clientID == -1) {
     throw "simxStart error";
   }
+  for (uint16_t i = 0; i < _maxRobots; ++i) {
+    _robots.push_back(Robot(_clientID));
+    usleep(100);
```
Comment (JeremyPouyet) [inline srcs/Simulation.cpp]: Why is there a usleep here ?

File: srcs/Robot.cpp [hunk 4]
```diff
@@ -45,9 +45,15 @@ Robot::Robot(simxInt clientID) :
   _clientID(clientID), _handlers(highestAction()), _id(_nextId) {
   ++_nextId;
   // here we throw but it's not a problem as this will later be done in another function
-  for (int i = 0; i < highestAction(); i++)
-    if (simxGetObjectHandle(_clientID, Robot::actions.at(i)._bodyPart, &_handlers[i], simx_opmode_blocking) != 0)
+  const char *id = std::to_string(_id).c_str();
```
Comment (JeremyPouyet) [inline srcs/Robot.cpp]: can't lines 48 to 53 be replaced with the use of std::string ?

File: includes/Robot.hh (hunk 2 above)
Comment (JeremyPouyet) [inline includes/Robot.hh]: in c++ you usually include `<cmath>` and not `<math.h>`

**Review by JeremyPouyet — COMMENTED** [review]
Just few questions

File: srcs/Robot.cpp [hunk 5]
```diff
@@ -66,7 +72,7 @@ simxInt		Robot::doActions(const dna_t &dna) const {
       continue;
     }
     cout << "Robot " << (int)_id << " executing " << actions.at(action)._description << endl;
-    ret = simxSetJointTargetPosition(_clientID, _handlers[action], Robot::actions.at(action)._intensity, simx_opmode_oneshot);
+    ret = simxSetJointTargetPosition(_clientID, _handlers[action], (Robot::actions.at(action)._intensity * M_PI) / 180, simx_opmode_blocking);
```
Comment (JeremyPouyet) [inline srcs/Robot.cpp]: should all values be multiplied by M_PI / 180 ?

File: srcs/Simulation.cpp (hunk 3 above)
Comment (1475963) [reply srcs/Simulation.cpp]: Sometimes there is an error thrown by the robot creation function, I tought that waiting between two api calls could help but no it doesn't

File: srcs/Robot.cpp (hunk 5 above)
Comment (1475963) [reply srcs/Robot.cpp]: yes

File: srcs/Simulation.cpp (hunk 3 above)
Comment (JeremyPouyet) [reply srcs/Simulation.cpp]: Can't you try to correct the problem before your push ? Or at least remove the usleep if it does'nt help

**Comment by jgburet** [comment]
Looks like we've been two working on that...

I also updated how the simulation was starting.
Indeed, it didn't make sense to create the population and whatever even before knowing if the connection to the remove VREP server was working.

Changes lead me to the Robot class where I updated & cleaned few things for more clarity.

File: srcs/Robot.cpp [hunk 6]
```diff
@@ -58,17 +69,25 @@ Robot::Robot(simxInt clientID) :
...
     if (action > highestAction()) {
       cerr << "Robot " << (int)_id << " can't execute " << (int)action << ": It does not exists!" << endl;
       continue;
     }
-    cout << "Robot " << (int)_id << " executing " << actions.at(action)._description << endl;
-    ret = simxSetJointTargetPosition(_clientID, _handlers[action], Robot::actions.at(action)._intensity, simx_opmode_oneshot);
-    if (ret != 0 && ret != 1)
-      return ret;
+    cout << "Robot " << (int)_id << " executing " << _movements.at(action).description << endl;
+    auto movementResult = simxSetJointTargetPosition(
```
Comment (JeremyPouyet) [inline srcs/Robot.cpp:78]: please do not use auto everywhere.

File: srcs/Robot.cpp [hunk 7]
```diff
@@ -2,53 +2,64 @@
...
-  for (int i = 0; i < highestAction(); i++)
-    if (simxGetObjectHandle(_clientID, Robot::actions.at(i)._bodyPart, &_handlers[i], simx_opmode_blocking) != 0)
-      throw "err";
-  cout << "Robot " << (int)_id << " created" << endl;
+Robot::Robot(simxInt clientID, robot_id_t id) :
+  _clientID(clientID),
+  _id(id) {
+
+  // Retrieving movements handlers
+  for (auto movement : _movements) {
```
Comment (JeremyPouyet) [inline srcs/Robot.cpp]: can be a little better with `for (const auto &movement : _movements)`

**Comment by JeremyPouyet** [comment]
@1475963 you have not added the 8 robots scenes.

File: srcs/Robot.cpp (hunk 6 above)
Comment (jgburet) [reply srcs/Robot.cpp:78]: `auto` is authorized when type is guessable or flagrant.
Here, the third parties library returns an `int` equal to `-1` on errors.

File: srcs/Robot.cpp (hunk 7 above)
Comment (jgburet) [reply srcs/Robot.cpp]: true

**Review by JeremyPouyet — APPROVED** [review]

**Pull request merged**